### PR TITLE
torchdynamo: add linear eltwise fusion kernel

### PR DESCRIPTION
**Pull request by chunyuan-w**
Support fusion of linear with:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu

### CODE REVIEW

**Comment by EikanWang** [comment]
@chunyuan-w , could you please add UT?

File: aten/src/ATen/native/mkldnn/OpContext.cpp [hunk 1]
```diff
@@ -7,6 +7,69 @@ namespace at {
...
+#define ATTR_FUNC(NAME)                              \
+  [](std::vector<c10::optional<at::Scalar>> scalars, \
+     c10::optional<std::string> algorithm) {         \
+    return ideep::attr_t::fuse_##NAME();             \
+  }
+
+AttrFunction attr_func_leaky_relu =
+    [](std::vector<c10::optional<at::Scalar>> scalars,
+       c10::optional<std::string> algorithm) {
+      TORCH_CHECK(scalars.size() == 1 && scalars[0].has_value(), "leaky_relu is expected to have one scalar input: negative_slope");
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/OpContext.cpp:19]: Is the aten as the same logic? Is the filter oneDNN fusion specific?

File: aten/src/ATen/native/mkldnn/OpContext.cpp [hunk 2]
```diff
@@ -7,6 +7,69 @@ namespace at {
...
+      auto alpha_value = scalars[0].value().to<float>();
+      return ideep::attr_t::fuse_relu(1.0, alpha_value);
+    };
+
+AttrFunction attr_func_hardtanh =
+    [](std::vector<c10::optional<at::Scalar>> scalars,
+       c10::optional<std::string> algorithm) {
+      TORCH_CHECK(
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/OpContext.cpp:27]: Ditto

File: aten/src/ATen/native/mkldnn/OpContext.cpp [hunk 3]
```diff
@@ -7,6 +7,69 @@ namespace at {
...
+      algorithm.has_value(),
+      "gelu is expected to have one str input: algorithm");
+  dnnl::algorithm gelu_type;
+  if (algorithm.value() == "none") {
+    gelu_type = dnnl::algorithm::eltwise_gelu_erf;
+  } else if (algorithm.value() == "tanh") {
+    gelu_type = dnnl::algorithm::eltwise_gelu_tanh;
+  } else {
+    TORCH_CHECK(
+        false, "Unsupported gelu algorithm: ", algorithm.value());
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/OpContext.cpp:54]: https://pytorch.org/docs/stable/generated/torch.nn.GELU.html

nn.GELu only supports "none" and "tanh". So I'd recommend using `TORCH_INTERNAL_ASSERT`

File: aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp [hunk 4]
```diff
@@ -42,6 +44,9 @@ TORCH_LIBRARY(mkldnn_prepacked, m) {
 
   m.def(TORCH_SELECTIVE_SCHEMA(
       "mkldnn_prepacked::conv2d_run(Tensor X, __torch__.torch.classes.mkldnn.ConvOpContext W_prepack) -> Tensor Y"));
+
+  m.def(TORCH_SELECTIVE_SCHEMA(
+    "mkldnn_prepacked::linear_eltwise(Tensor X, Tensor W, Tensor? B, str attr, Scalar?[] scalars, str? algorithm) -> Tensor Y"));
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp:49]: Is this schema general enough to cover all the elementwise operations regardless of its parameter number? Seems like yes as the `scalars` is a vector, but just confirm with you.

**Review by EikanWang — CHANGES_REQUESTED** [review]
Please add UT.

File: aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp (hunk 4 above)
Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp:49]: Yes, this schema covers all the targeted elementwise operators.

**Comment by chunyuan-w** [comment]
UTs are added. Lintrunner installation fails in the CI due to network issue.

File: aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp (hunk 4 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp:49]: it is confused with using mkldnn_prepacked for this op, it does not use prepack weight. also same as LinearPreapck file name.

File: aten/src/ATen/native/mkldnn/OpContext.cpp (hunk 3 above)
Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/OpContext.cpp]: Converted to `TORCH_INTERNAL_ASSERT`

**Comment by EikanWang** [comment]
@chunyuan-w , please check the failed CI cases.

File: aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp (hunk 4 above)
Comment (chunyuan-w) [reply aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp]: The OP is renamed to `torch.ops.mkldnn._linear_pointwise`.

**Comment by chunyuan-w** [comment]
> @chunyuan-w , please check the failed CI cases.

@EikanWang  I've merged the recent master and all CIs have passed.

**Comment by EikanWang** [comment]
LGTM

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
@frank-wei , this PR is to serve TorchInductor fusion and mkldnn-specific. But I could not merge it. Could you help review it? And I wonder if you can help to land it as well if the PR is okay with you.

**Review by jansel — APPROVED** [review]

**Comment by EikanWang** [comment]
@chunyuan-w ，please sign the EasyCLA and rebase this PR.

**Comment by chunyuan-w** [comment]
> @chunyuan-w ，please sign the EasyCLA and rebase this PR.

EasyCLA is signed. I've merged the recent master and all CIs have passed.
